File: DoublyLinkedList/DoublyLinkedList.py
"""
Doubly linked list implementation with basic operations.
Each node contains data and references to both previous and next nodes.
"""


class Node:
    """Node class for doubly linked list."""
    def __init__(self, data):
        self.data = data
        self.prev = None
        self.next = None


class DoublyLinkedList:
    """Doubly linked list with comprehensive operations."""

    def __init__(self):
        self.head = None
        self.tail = None
        self.size = 0

    def is_empty(self):
        """Check if list is empty. O(1) operation."""
        return self.head is None

    def get_size(self):
        """Return number of elements in list. O(1) operation."""
        return self.size

    def insert_at_beginning(self, data):
        """Insert element at the beginning of the list. O(1) operation."""
        new_node = Node(data)
        if self.is_empty():
            self.head = self.tail = new_node
        else:
            new_node.next = self.head
            self.head.prev = new_node
            self.head = new_node
        self.size += 1

    def insert_at_end(self, data):
        """Insert element at the end of the list. O(1) operation."""
        new_node = Node(data)
        if self.is_empty():
            self.head = self.tail = new_node
        else:
            new_node.prev = self.tail
            self.tail.next = new_node
            self.tail = new_node
        self.size += 1

    def insert_at_position(self, data, position):
        """Insert element at specified position. O(n) operation."""
        if position < 0 or position > self.size:
            raise ValueError("Invalid position")

        if position == 0:
            self.insert_at_beginning(data)
            return

        if position == self.size:
            self.insert_at_end(data)
            return

        new_node = Node(data)
        current = self.head
        for _ in range(position - 1):
            current = current.next

        new_node.next = current.next
        new_node.prev = current
        current.next.prev = new_node
        current.next = new_node
        self.size += 1

    def delete_from_beginning(self):
        """Delete element from the beginning of the list. O(1) operation."""
        if self.is_empty():
            return None

        deleted_data = self.head.data
        if self.head == self.tail:
            self.head = self.tail = None
        else:
            self.head = self.head.next
            self.head.prev = None
        self.size -= 1
        return deleted_data

    def delete_from_end(self):
        """Delete element from the end of the list. O(1) operation."""
        if self.is_empty():
            return None

        deleted_data = self.tail.data
        if self.head == self.tail:
            self.head = self.tail = None
        else:
            self.tail = self.tail.prev
            self.tail.next = None
        self.size -= 1
        return deleted_data

    def delete_by_value(self, value):
        """Delete first occurrence of value. O(n) operation."""
        if self.is_empty():
            return False

        current = self.head
        while current:
            if current.data == value:
                if current == self.head:
                    self.delete_from_beginning()
                elif current == self.tail:
                    self.delete_from_end()
                else:
                    current.prev.next = current.next
                    current.next.prev = current.prev
                    self.size -= 1
                return True
            current = current.next
        return False

    def search(self, value):
        """Search for value in the list. O(n) operation."""
        current = self.head
        position = 0
        while current:
            if current.data == value:
                return position
            current = current.next
            position += 1
        return -1

    def display_forward(self):
        """Display all elements from head to tail."""
        if self.is_empty():
            print("List is empty")
            return

        current = self.head
        elements = []
        while current:
            elements.append(str(current.data))
            current = current.next
        print(" -> ".join(elements))

    def display_backward(self):
        """Display all elements from tail to head."""
        if self.is_empty():
            print("List is empty")
            return

        current = self.tail
        elements = []
        while current:
            elements.append(str(current.data))
            current = current.prev
        print(" <- ".join(elements))


def main():
    """Demonstrates doubly linked list operations."""
    import time
    import sys

    start_time = time.time()
    memory_before = sys.getsizeof([])  # Approximate baseline

    # Create doubly linked list instance
    dll = DoublyLinkedList()

    # Insert elements
    dll.insert_at_end(10)
    dll.insert_at_end(20)
    dll.insert_at_beginning(5)
    dll.insert_at_position(15, 2)

    print("List after insertions (forward):")
    dll.display_forward()

    print("\nList after insertions (backward):")
    dll.display_backward()

    # Search operations
    search_values = [15, 25]
    print("\nSearch results:")
    for value in search_values:
        position = dll.search(value)
        if position != -1:
            print(f"Value {value} found at position {position}")
        else:
            print(f"Value {value} not found")

    # Delete operations
    dll.delete_by_value(20)
    print("\nList after deleting 20 (forward):")
    dll.display_forward()

    dll.delete_from_beginning()
    print("\nList after deleting from beginning (forward):")
    dll.display_forward()

    dll.delete_from_end()
    print("\nList after deleting from end (forward):")
    dll.display_forward()

    end_time = time.time()
    memory_after = sys.getsizeof([])  # Approximate after

    # Performance statistics
    execution_time = (end_time - start_time) * 1000  # Convert to ms
    memory_used = memory_after - memory_before

    print("\n--- Performance Statistics ---")
    print(f"Execution time: {execution_time:.4f} ms")
    print(f"Memory usage: {memory_used if memory_used > 0 else 'negligible'} bytes")
    print(f"Final list size: {dll.get_size()}")
    print("\nTime Complexity:")
    print("  - Insert at beginning: O(1)")
    print("  - Insert at end: O(1)")
    print("  - Insert at position: O(n)")
    print("  - Delete from beginning: O(1)")
    print("  - Delete from end: O(1)")
    print("  - Delete by value: O(n)")
    print("  - Search: O(n)")
    print("  - Get size: O(1)")


if __name__ == "__main__":
    main()

File: DoublyLinkedList/DoublyLinkedList.go
// Doubly linked list implementation with basic operations.
// Each node contains data and pointers to both previous and next nodes.
package main

import (
	"fmt"
	"time"
	"unsafe"
)

// Node represents a node in the doubly linked list
type Node[T comparable] struct {
	data T
	prev *Node[T]
	next *Node[T]
}

// NewNode creates a new node with the given data
func NewNode[T comparable](data T) *Node[T] {
	return &Node[T]{data: data, prev: nil, next: nil}
}

// DoublyLinkedList represents a doubly linked list
type DoublyLinkedList[T comparable] struct {
	head *Node[T]
	tail *Node[T]
	size int
}

// NewDoublyLinkedList creates a new empty doubly linked list
func NewDoublyLinkedList[T comparable]() *DoublyLinkedList[T] {
	return &DoublyLinkedList[T]{head: nil, tail: nil, size: 0}
}

// IsEmpty checks if the list is empty. O(1) operation.
func (dll *DoublyLinkedList[T]) IsEmpty() bool {
	return dll.head == nil
}

// GetSize returns the number of elements in the list. O(1) operation.
func (dll *DoublyLinkedList[T]) GetSize() int {
	return dll.size
}

// InsertAtBeginning inserts an element at the beginning of the list. O(1) operation.
func (dll *DoublyLinkedList[T]) InsertAtBeginning(data T) {
	newNode := NewNode(data)
	if dll.IsEmpty() {
		dll.head = newNode
		dll.tail = newNode
	} else {
		newNode.next = dll.head
		dll.head.prev = newNode
		dll.head = newNode
	}
	dll.size++
}

// InsertAtEnd inserts an element at the end of the list. O(1) operation.
func (dll *DoublyLinkedList[T]) InsertAtEnd(data T) {
	newNode := NewNode(data)
	if dll.IsEmpty() {
		dll.head = newNode
		dll.tail = newNode
	} else {
		newNode.prev = dll.tail
		dll.tail.next = newNode
		dll.tail = newNode
	}
	dll.size++
}

// InsertAtPosition inserts an element at the specified position. O(n) operation.
func (dll *DoublyLinkedList[T]) InsertAtPosition(data T, position int) {
	if position < 0 || position > dll.size {
		panic("Invalid position")
	}

	if position == 0 {
		dll.InsertAtBeginning(data)
		return
	}

	if position == dll.size {
		dll.InsertAtEnd(data)
		return
	}

	newNode := NewNode(data)
	current := dll.head
	for i := 0; i < position-1; i++ {
		current = current.next
	}

	newNode.next = current.next
	newNode.prev = current
	current.next.prev = newNode
	current.next = newNode
	dll.size++
}

// DeleteFromBeginning deletes the element from the beginning of the list. O(1) operation.
func (dll *DoublyLinkedList[T]) DeleteFromBeginning() T {
	if dll.IsEmpty() {
		panic("List is empty")
	}

	deletedData := dll.head.data
	if dll.head == dll.tail {
		dll.head = nil
		dll.tail = nil
	} else {
		dll.head = dll.head.next
		dll.head.prev = nil
	}
	dll.size--
	return deletedData
}

// DeleteFromEnd deletes the element from the end of the list. O(1) operation.
func (dll *DoublyLinkedList[T]) DeleteFromEnd() T {
	if dll.IsEmpty() {
		panic("List is empty")
	}

	deletedData := dll.tail.data
	if dll.head == dll.tail {
		dll.head = nil
		dll.tail = nil
	} else {
		dll.tail = dll.tail.prev
		dll.tail.next = nil
	}
	dll.size--
	return deletedData
}

// DeleteByValue deletes the first occurrence of the value. O(n) operation.
func (dll *DoublyLinkedList[T]) DeleteByValue(value T) bool {
	if dll.IsEmpty() {
		return false
	}

	current := dll.head
	for current != nil {
		if current.data == value {
			if current == dll.head {
				dll.DeleteFromBeginning()
			} else if current == dll.tail {
				dll.DeleteFromEnd()
			} else {
				current.prev.next = current.next
				current.next.prev = current.prev
				dll.size--
			}
			return true
		}
		current = current.next
	}
	return false
}

// Search searches for a value in the list. O(n) operation.
func (dll *DoublyLinkedList[T]) Search(value T) int {
	current := dll.head
	position := 0
	for current != nil {
		if current.data == value {
			return position
		}
		current = current.next
		position++
	}
	return -1
}

// DisplayForward prints all elements from head to tail.
func (dll *DoublyLinkedList[T]) DisplayForward() {
	if dll.IsEmpty() {
		fmt.Println("List is empty")
		return
	}

	current := dll.head
	for current != nil {
		fmt.Print(current.data)
		if current.next != nil {
			fmt.Print(" -> ")
		}
		current = current.next
	}
	fmt.Println()
}

// DisplayBackward prints all elements from tail to head.
func (dll *DoublyLinkedList[T]) DisplayBackward() {
	if dll.IsEmpty() {
		fmt.Println("List is empty")
		return
	}

	current := dll.tail
	for current != nil {
		fmt.Print(current.data)
		if current.prev != nil {
			fmt.Print(" <- ")
		}
		current = current.prev
	}
	fmt.Println()
}

// main demonstrates doubly linked list operations.
// Performs insertions, deletions, and searches on the list.
func main() {
	startTime := time.Now()

	// Create doubly linked list instance
	dll := NewDoublyLinkedList[int]()

	// Insert elements
	dll.InsertAtEnd(10)
	dll.InsertAtEnd(20)
	dll.InsertAtBeginning(5)
	dll.InsertAtPosition(15, 2)

	fmt.Println("List after insertions (forward):")
	dll.DisplayForward()

	fmt.Println("\nList after insertions (backward):")
	dll.DisplayBackward()

	// Search operations
	searchValues := []int{15, 25}
	fmt.Println("\nSearch results:")
	for _, value := range searchValues {
		position := dll.Search(value)
		if position != -1 {
			fmt.Printf("Value %d found at position %d\n", value, position)
		} else {
			fmt.Printf("Value %d not found\n", value)
		}
	}

	// Delete operations
	dll.DeleteByValue(20)
	fmt.Println("\nList after deleting 20 (forward):")
	dll.DisplayForward()

	dll.DeleteFromBeginning()
	fmt.Println("\nList after deleting from beginning (forward):")
	dll.DisplayForward()

	dll.DeleteFromEnd()
	fmt.Println("\nList after deleting from end (forward):")
	dll.DisplayForward()

	endTime := time.Now()
	executionTime := endTime.Sub(startTime)

	// Calculate memory usage (approximate)
	var dummyNode Node[int]
	nodeSize := unsafe.Sizeof(dummyNode)
	memoryUsed := uintptr(dll.size) * nodeSize

	// Performance statistics
	fmt.Println("\n--- Performance Statistics ---")
	fmt.Printf("Execution time: %.4f ms\n", float64(executionTime.Nanoseconds())/1e6)
	fmt.Printf("Memory usage: %d bytes\n", memoryUsed)
	fmt.Printf("Final list size: %d\n", dll.GetSize())
	fmt.Println("\nTime Complexity:")
	fmt.Println("  - Insert at beginning: O(1)")
	fmt.Println("  - Insert at end: O(1)")
	fmt.Println("  - Insert at position: O(n)")
	fmt.Println("  - Delete from beginning: O(1)")
	fmt.Println("  - Delete from end: O(1)")
	fmt.Println("  - Delete by value: O(n)")
	fmt.Println("  - Search: O(n)")
	fmt.Println("  - Get size: O(1)")
}